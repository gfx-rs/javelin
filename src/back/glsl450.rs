use crate::{
    back::glsl_common::{write_type, Error, StatementBuilder},
    Binding, BuiltIn, FastHashMap, Module, TypeInner,
};
use std::{
    fmt::{self},
    io::Write,
};

pub fn write(module: &Module, out: &mut impl Write) -> Result<(), Error> {
    writeln!(out, "#version 450 core")?;

    let mut counter = 0;
    let mut names = FastHashMap::default();

    let mut namer = |name: Option<&String>| {
        if let Some(name) = name {
            names.insert(name.clone(), ());
            name.clone()
        } else {
            counter += 1;
            while names.get(&format!("_{}", counter)).is_some() {
                counter += 1;
            }
            format!("_{}", counter)
        }
    };

    let mut structs = FastHashMap::default();

    // Do a first pass to collect names
    for (handle, ty) in module.types.iter() {
        match ty.inner {
            TypeInner::Struct { ref members } => {
                let name = namer(ty.name.as_ref());
                let mut idxs = Vec::new();

                for member in members {
                    idxs.push(namer(member.name.as_ref()))
                }

                structs.insert(handle, (name, idxs));
            }
            _ => continue,
        }
    }

    // Do a second pass to build the structs
    // TODO: glsl is order dependent so we need to build structs in order
    for (handle, ty) in module.types.iter() {
        match ty.inner {
            TypeInner::Struct { ref members } => {
                let (name, idxs) = structs.get(&handle).unwrap();

                writeln!(out, "struct {} {{", name)?;
                for (member, name) in members.iter().zip(idxs.iter()) {
                    writeln!(
                        out,
                        "   {} {}",
                        write_type(member.ty, &module.types, &structs)?,
                        name
                    )?;
                }
                writeln!(out, "}};")?;
            }
            _ => continue,
        }
    }

    let mut globals_lookup = FastHashMap::default();

    for (handle, global) in module.global_variables.iter() {
        if let Some(Binding::BuiltIn(built_in)) = global.binding {
            let semantic = match built_in {
                BuiltIn::Position => "gl_position",
                BuiltIn::GlobalInvocationId => "gl_GlobalInvocationID",
                BuiltIn::BaseInstance => "gl_BaseInstance",
                BuiltIn::BaseVertex => "gl_BaseVertex",
                BuiltIn::ClipDistance => "gl_ClipDistance",
                BuiltIn::InstanceIndex => "gl_InstanceIndex",
                BuiltIn::VertexIndex => "gl_VertexIndex",
                BuiltIn::PointSize => "gl_PointSize",
                BuiltIn::FragCoord => "gl_FragCoord",
                BuiltIn::FrontFacing => "gl_FrontFacing",
                BuiltIn::SampleIndex => "gl_SampleID",
                BuiltIn::FragDepth => "gl_FragDepth",
                BuiltIn::LocalInvocationId => "gl_LocalInvocationID",
                BuiltIn::LocalInvocationIndex => "gl_LocalInvocationIndex",
                BuiltIn::WorkGroupId => "gl_WorkGroupID",
            };

            globals_lookup.insert(handle, String::from(semantic));
            continue;
        }

        if let Some(ref binding) = global.binding {
            write!(out, "layout({}) ", binding.writer())?;
        }

        let name = namer(global.name.as_ref());

        writeln!(
            out,
            "{}{} {};",
            global.class.write_glsl(),
            write_type(global.ty, &module.types, &structs)?,
            name
        )?;

        globals_lookup.insert(handle, name);
    }

    let mut functions = FastHashMap::default();

    // Do a first pass to collect names
    for (handle, func) in module.functions.iter() {
        functions.insert(handle, namer(func.name.as_ref()));
    }

    let mut types = module.types.clone();
    let mut typifier = crate::proc::Typifier::new();

    // TODO: glsl is order dependent so we need to build functions in order
    for (handle, func) in module.functions.iter() {
        let name = functions.get(&handle).unwrap();

        writeln!(
            out,
            "{} {}({}) {{",
            func.return_type
                .map_or(Ok(String::from("void")), |ty| write_type(
                    ty, &types, &structs
                ))?,
            name,
            func.parameter_types
                .iter()
                .map(|ty| write_type(*ty, &types, &structs))
                .collect::<Result<Vec<_>, _>>()?
                .join(","),
        )?;

        let mut builder = StatementBuilder {
            functions: &functions,
            globals: &globals_lookup,
            locals_lookup: &func
                .local_variables
                .iter()
                .map(|(handle, local)| (handle, namer(local.name.as_ref())))
                .collect(),
            structs: &structs,
            args: &func
                .parameter_types
                .iter()
                .enumerate()
                .map(|(pos, _)| (pos as u32, namer(None)))
                .collect(),
            expressions: &func.expressions,
            types: &mut types,
            locals: &func.local_variables,
            typifier: &mut typifier,
        };

        for sta in func.body.iter() {
            writeln!(out, "{}", sta.write_glsl(module, &mut builder)?)?;
        }

        writeln!(out, "}}")?;
    }

    Ok(())
}

struct BindingWriter<'a> {
    inner: &'a Binding,
}

impl Binding {
    fn writer<'a>(&'a self) -> BindingWriter<'a> {
        BindingWriter { inner: self }
    }
}

impl<'a> fmt::Display for BindingWriter<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self.inner {
            Binding::BuiltIn(_) => write!(f, ""), // Ignore because they are variables with a predefined name
            Binding::Location(location) => write!(f, "location={}", location),
            Binding::Descriptor { set, binding } => write!(f, "set={},binding={}", set, binding),
        }
    }
}
