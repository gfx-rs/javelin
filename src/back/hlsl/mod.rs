mod keywords;
mod writer;

use std::fmt::Error as FmtError;
use thiserror::Error;

pub use writer::Writer;

#[derive(Debug, Copy, Clone, PartialEq, PartialOrd)]
pub struct ShaderModel(u16);

impl Default for ShaderModel {
    fn default() -> Self {
        ShaderModel(50)
    }
}

#[derive(Error, Debug)]
pub enum Error {
    #[error(transparent)]
    IoError(#[from] FmtError),
    #[error("BuiltIn {0:?} is not supported")]
    UnsupportedShaderModel(ShaderModel),
    #[error("A scalar with an unsupported width was requested: {0:?} {1:?}")]
    UnsupportedScalar(crate::ScalarKind, crate::Bytes),
    #[error("BuiltIn {0:?} is not supported")]
    UnsupportedBuiltIn(crate::BuiltIn),
    #[error("{0}")]
    Unimplemented(String), // TODO: Error used only during development
}

pub fn write_string(
    module: &crate::Module,
    info: &crate::valid::ModuleInfo,
    shader_model: ShaderModel,
) -> Result<String, Error> {
    let mut w = Writer::new(String::new(), shader_model);
    w.write(module, info)?;
    let output = w.finish();
    Ok(output)
}
