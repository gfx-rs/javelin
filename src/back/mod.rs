//! Functions which export shader modules into binary and text formats.

#[cfg(feature = "glsl-out")]
pub mod glsl;
#[cfg(feature = "msl-out")]
pub mod msl;
#[cfg(feature = "spv-out")]
pub mod spv;

impl crate::Expression {
    /// Returns the ref count, upon reaching which this expression
    /// should be considered for baking.
    #[allow(dead_code)]
    fn bake_ref_count(&self) -> usize {
        match *self {
            // accesses are never cached, only loads are
            crate::Expression::Access { .. } | crate::Expression::AccessIndex { .. } => !0,
            // image operations look better when isolated
            crate::Expression::ImageSample { .. } | crate::Expression::ImageLoad { .. } => 1,
            // cache expressions that are referenced multiple times
            _ => 2,
        }
    }
}
