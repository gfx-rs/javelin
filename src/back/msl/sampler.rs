#[derive(Clone, Debug, PartialEq)]
pub enum Coord {
    Normalized,
    Pixel,
}

impl Default for Coord {
    fn default() -> Self {
        Self::Normalized
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum Address {
    Repeat,
    MirroredRepeat,
    ClampToEdge,
    ClampToZero,
    ClampToBorder,
}

impl Default for Address {
    fn default() -> Self {
        Self::ClampToEdge
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum BorderColor {
    TransparentBlack,
    OpaqueBlack,
    OpaqueWhite,
}

impl Default for BorderColor {
    fn default() -> Self {
        Self::TransparentBlack
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum Filter {
    Nearest,
    Linear,
}

impl Default for Filter {
    fn default() -> Self {
        Self::Nearest
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum CompareFunc {
    Never,
    Less,
    LessEqual,
    Greater,
    GreaterEqual,
    Equal,
    NotEqual,
    Always,
}

impl Default for CompareFunc {
    fn default() -> Self {
        Self::Never
    }
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct InlineSampler {
    pub coord: Coord,
    pub address: [Address; 3],
    pub border_color: BorderColor,
    pub mag_filter: Filter,
    pub min_filter: Filter,
    pub mip_filter: Option<Filter>,
    pub compare_func: CompareFunc,
}
