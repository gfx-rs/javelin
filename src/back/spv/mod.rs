mod layout;
pub mod parser;

use spirv::*;

struct Module {
    physical_layout: PhysicalLayout,
    logical_layout: LogicalLayout,
}

impl Module {
    fn new(header: &crate::Header) -> Self {
        Module {
            physical_layout: PhysicalLayout::new(header, 0x0u32),
            logical_layout: LogicalLayout::new(),
        }
    }
}

struct PhysicalLayout {
    magic_number: Word,
    version: Word,
    generator: Word,
    bound: Word,
    instruction_schema: Word,
}

impl PhysicalLayout {
    fn new(header: &crate::Header, instruction_schema: Word) -> Self {
        let version: Word = (0x0u32 << 24)
            | ((header.version.0 as u32) << 16)
            | ((header.version.1 as u32) << 8)
            | header.version.2 as u32;

        PhysicalLayout {
            magic_number: MAGIC_NUMBER,
            version,
            generator: header.generator,
            bound: 0,
            instruction_schema,
        }
    }

    fn in_words(&self) -> Vec<Word> {
        let mut words: Vec<Word> = vec![];

        words.push(self.magic_number);
        words.push(self.version);
        words.push(self.generator);

        match self.bound {
            0 => panic!("Bound cannot be 0"),
            _ => words.push(self.bound),
        }

        words.push(self.instruction_schema);
        words
    }
}

pub(crate) struct LogicalLayout {
    capabilities: Vec<Word>,
    extensions: Vec<Word>,
    ext_inst_imports: Vec<Word>,
    memory_model: Vec<Word>,
    entry_points: Vec<Word>,
    execution_modes: Vec<Word>,
    debugs: Vec<Word>,
    annotations: Vec<Word>,
    type_declarations: Vec<Word>,
    constants: Vec<Word>,
    global_variables: Vec<Word>,
    function_declarations: Vec<Word>,
    function_definitions: Vec<Word>,
}

impl LogicalLayout {
    fn new() -> Self {
        LogicalLayout {
            capabilities: vec![],
            extensions: vec![],
            ext_inst_imports: vec![],
            memory_model: vec![],
            entry_points: vec![],
            execution_modes: vec![],
            debugs: vec![],
            annotations: vec![],
            type_declarations: vec![],
            constants: vec![],
            global_variables: vec![],
            function_declarations: vec![],
            function_definitions: vec![],
        }
    }

    fn in_words(&self) -> Vec<u32> {
        let mut words: Vec<Word> = vec![];

        for capability in self.capabilities.iter() {
            words.push(*capability);
        }

        for extension in self.extensions.iter() {
            words.push(*extension);
        }

        for ext_inst_import in self.ext_inst_imports.iter() {
            words.push(*ext_inst_import);
        }

        for memory_model in self.memory_model.iter() {
            words.push(*memory_model);
        }

        for entry_point in self.entry_points.iter() {
            words.push(*entry_point);
        }

        for execution_mode in self.execution_modes.iter() {
            words.push(*execution_mode);
        }

        for debug in self.debugs.iter() {
            words.push(*debug);
        }

        for annotation in self.annotations.iter() {
            words.push(*annotation);
        }

        for type_declaration in self.type_declarations.iter() {
            words.push(*type_declaration);
        }

        for constants in self.constants.iter() {
            words.push(*constants);
        }

        for global_variables in self.global_variables.iter() {
            words.push(*global_variables);
        }

        for function_declaration in self.function_declarations.iter() {
            words.push(*function_declaration);
        }

        for function_definition in self.function_definitions.iter() {
            words.push(*function_definition);
        }

        words
    }
}

struct Instruction {
    op: Op,
    wc: u32,
    type_id: Option<Word>,
    result_id: Option<Word>,
    operands: Vec<Word>,
}

impl Instruction {
    fn new(op: Op) -> Self {
        Instruction {
            op,
            wc: 1, // Always start at 1 for the first word (OP + WC),
            type_id: None,
            result_id: None,
            operands: vec![],
        }
    }

    fn set_type(&mut self, id: Word) {
        assert!(self.type_id.is_none(), "Type can only be set once");
        self.type_id = Some(id);
        self.wc += 1;
    }

    fn set_result(&mut self, id: Word) {
        assert!(self.result_id.is_none(), "Result can only be set once");
        self.result_id = Some(id);
        self.wc += 1;
    }

    fn add_operand(&mut self, operand: Word) {
        self.operands.push(operand);
        self.wc += 1;
    }

    fn add_operands(&mut self, operands: Vec<Word>) {
        for operand in operands {
            self.add_operand(operand);
        }
    }

    fn to_words(&self) -> Vec<Word> {
        let mut words = Vec::with_capacity(self.wc as usize);

        let wc_op = (self.wc << 16 | self.op as u32) as u32;

        words.push(wc_op);

        if self.type_id.is_some() {
            words.push(self.type_id.unwrap());
        }

        if self.result_id.is_some() {
            words.push(self.result_id.unwrap());
        }

        for operand in self.operands.iter() {
            words.push(*operand);
        }

        words
    }
}
