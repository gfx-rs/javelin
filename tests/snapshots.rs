use serde::Deserialize;

#[derive(Hash, PartialEq, Eq, Deserialize)]
enum Stage {
    Vertex,
    Fragment,
    Compute,
}

#[derive(Hash, PartialEq, Eq, Deserialize)]
struct BindSource {
    stage: Stage,
    group: u32,
    binding: u32,
}

#[derive(Deserialize)]
struct BindTarget {
    #[serde(default)]
    buffer: Option<u8>,
    #[serde(default)]
    texture: Option<u8>,
    #[serde(default)]
    sampler: Option<u8>,
    #[serde(default)]
    mutable: bool,
}

#[derive(Default, Deserialize)]
struct Parameters {
    #[serde(default)]
    spv_flow_dump_prefix: String,
    #[cfg_attr(not(feature = "spv-out"), allow(dead_code))]
    spv_capabilities: naga::FastHashSet<spirv::Capability>,
    #[cfg_attr(not(feature = "msl-out"), allow(dead_code))]
    mtl_bindings: naga::FastHashMap<BindSource, BindTarget>,
}

#[cfg(feature = "msl-out")]
fn check_output_msl(module: &naga::Module, name: &str, params: &Parameters) {
    use naga::back::msl;
    let mut binding_map = msl::BindingMap::default();
    for (key, value) in params.mtl_bindings.iter() {
        binding_map.insert(
            msl::BindSource {
                stage: match key.stage {
                    Stage::Vertex => naga::ShaderStage::Vertex,
                    Stage::Fragment => naga::ShaderStage::Fragment,
                    Stage::Compute => naga::ShaderStage::Compute,
                },
                group: key.group,
                binding: key.binding,
            },
            msl::BindTarget {
                buffer: value.buffer,
                texture: value.texture,
                sampler: value.sampler,
                mutable: value.mutable,
            },
        );
    }
    let options = msl::Options {
        lang_version: (1, 0),
        spirv_cross_compatibility: false,
        binding_map,
    };
    let (msl, _) = msl::write_string(&module, &options).unwrap();
    insta::assert_snapshot!(format!("{}{}", name, ".msl"), msl);
}

#[cfg(feature = "spv-out")]
fn check_output_spv(module: &naga::Module, name: &str, params: &Parameters) {
    use naga::back::spv;
    use rspirv::binary::Disassemble;
    let spv = spv::write_vec(
        &module,
        spv::WriterFlags::NONE,
        params.spv_capabilities.clone(),
    )
    .unwrap();
    let dis = rspirv::dr::load_words(spv)
        .expect("Produced invalid SPIR-V")
        .disassemble();
    insta::assert_snapshot!(format!("{}{}", name, ".spvasm"), dis);
}

#[cfg(feature = "wgsl-in")]
fn convert_wgsl(name: &str) {
    let params =
        match std::fs::read_to_string(format!("tests/snapshots/in/{}{}", name, ".param.ron")) {
            Ok(string) => ron::de::from_str(&string).expect("Couldn't find param file"),
            Err(_) => Parameters::default(),
        };

    let module = naga::front::wgsl::parse_str(
        &std::fs::read_to_string(format!("tests/snapshots/in/{}{}", name, ".wgsl"))
            .expect("Couldn't find wgsl file"),
    )
    .unwrap();
    naga::proc::Validator::new().validate(&module).unwrap();

    #[cfg(feature = "msl-out")]
    check_output_msl(&module, name, &params);
    #[cfg(feature = "spv-out")]
    check_output_spv(&module, name, &params);
}

#[cfg(feature = "wgsl-in")]
#[test]
fn converts_wgsl_quad() {
    convert_wgsl("quad");
}
