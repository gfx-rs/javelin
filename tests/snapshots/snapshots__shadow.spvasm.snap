---
source: tests/snapshots.rs
expression: dis
---
; SPIR-V
; Version: 1.0
; Generator: rspirv
; Bound: 221
OpCapability Shader
%1 = OpExtInstImport "GLSL.std.450"
OpMemoryModel Logical GLSL450
OpEntryPoint Fragment %64 "fs_main" %114 %111 %217
OpExecutionMode %64 OriginUpperLeft
OpDecorate %27 DescriptorSet 0
OpDecorate %27 Binding 2
OpDecorate %32 DescriptorSet 0
OpDecorate %32 Binding 3
OpDecorate %76 Block
OpMemberDecorate %76 0 Offset 0
OpDecorate %77 DescriptorSet 0
OpDecorate %77 Binding 0
OpMemberDecorate %97 0 Offset 0
OpMemberDecorate %97 0 ColMajor
OpMemberDecorate %97 0 MatrixStride 16
OpMemberDecorate %97 1 Offset 64
OpMemberDecorate %97 2 Offset 80
OpDecorate %99 ArrayStride 96
OpDecorate %101 BufferBlock
OpMemberDecorate %101 0 Offset 0
OpMemberDecorate %101 0 NonWritable
OpDecorate %102 DescriptorSet 0
OpDecorate %102 Binding 1
OpDecorate %111 Location 1
OpDecorate %114 Location 0
OpDecorate %217 Location 0
%3 = OpTypeFloat 32
%2 = OpConstant  %3  0.0
%4 = OpConstant  %3  1.0
%5 = OpConstant  %3  0.5
%6 = OpConstant  %3  -0.5
%7 = OpConstant  %3  0.05
%9 = OpTypeVector %3 3
%8 = OpConstantComposite  %9  %7 %7 %7
%11 = OpTypeInt 32 0
%10 = OpConstant  %11  10
%12 = OpConstant  %11  0
%13 = OpConstant  %11  1
%16 = OpTypeVector %3 4
%18 = OpTypeFunction %3 %11 %16
%20 = OpTypeBool
%26 = OpTypeImage %3 2D 1 1 0 1 Unknown
%28 = OpTypePointer UniformConstant %26
%27 = OpVariable  %28  UniformConstant
%30 = OpTypeSampledImage %26
%31 = OpTypeSampler
%33 = OpTypePointer UniformConstant %31
%32 = OpVariable  %33  UniformConstant
%35 = OpTypeVector %3 2
%49 = OpTypeInt 32 1
%58 = OpConstant  %3  0.0
%60 = OpTypePointer Function %9
%62 = OpTypePointer Function %11
%63 = OpTypeVoid
%65 = OpTypeFunction %63
%74 = OpTypeVector %11 4
%76 = OpTypeStruct %74
%78 = OpTypePointer Uniform %76
%77 = OpVariable  %78  Uniform
%79 = OpTypePointer Uniform %74
%80 = OpConstant  %49  0
%81 = OpTypePointer Uniform %11
%82 = OpConstant  %49  0
%95 = OpTypeMatrix %16 4
%97 = OpTypeStruct %95 %16 %16
%99 = OpTypeRuntimeArray %97
%101 = OpTypeStruct %99
%103 = OpTypePointer Uniform %101
%102 = OpVariable  %103  Uniform
%104 = OpTypePointer Uniform %99
%105 = OpConstant  %49  0
%107 = OpTypePointer Uniform %97
%108 = OpTypePointer Uniform %95
%109 = OpConstant  %49  0
%112 = OpTypePointer Input %16
%111 = OpVariable  %112  Input
%115 = OpTypePointer Input %9
%114 = OpVariable  %115  Input
%123 = OpTypePointer Uniform %99
%124 = OpConstant  %49  0
%126 = OpTypePointer Uniform %97
%127 = OpTypePointer Uniform %16
%128 = OpConstant  %49  1
%129 = OpTypePointer Uniform %3
%130 = OpConstant  %49  0
%136 = OpTypePointer Uniform %99
%137 = OpConstant  %49  0
%139 = OpTypePointer Uniform %97
%140 = OpTypePointer Uniform %16
%141 = OpConstant  %49  1
%142 = OpTypePointer Uniform %3
%143 = OpConstant  %49  1
%149 = OpTypePointer Uniform %99
%150 = OpConstant  %49  0
%152 = OpTypePointer Uniform %97
%153 = OpTypePointer Uniform %16
%154 = OpConstant  %49  1
%155 = OpTypePointer Uniform %3
%156 = OpConstant  %49  2
%160 = OpTypePointer Input %3
%161 = OpConstant  %49  0
%164 = OpTypePointer Input %3
%165 = OpConstant  %49  1
%168 = OpTypePointer Input %3
%169 = OpConstant  %49  2
%179 = OpTypePointer Uniform %99
%180 = OpConstant  %49  0
%182 = OpTypePointer Uniform %97
%183 = OpTypePointer Uniform %16
%184 = OpConstant  %49  2
%185 = OpTypePointer Uniform %3
%186 = OpConstant  %49  0
%192 = OpTypePointer Uniform %99
%193 = OpConstant  %49  0
%195 = OpTypePointer Uniform %97
%196 = OpTypePointer Uniform %16
%197 = OpConstant  %49  2
%198 = OpTypePointer Uniform %3
%199 = OpConstant  %49  1
%205 = OpTypePointer Uniform %99
%206 = OpConstant  %49  0
%208 = OpTypePointer Uniform %97
%209 = OpTypePointer Uniform %16
%210 = OpConstant  %49  2
%211 = OpTypePointer Uniform %3
%212 = OpConstant  %49  2
%218 = OpTypePointer Output %16
%217 = OpVariable  %218  Output
%17 = OpFunction  %3  None %18
%14 = OpFunctionParameter  %11
%15 = OpFunctionParameter  %16
%19 = OpLabel
%22 = OpCompositeExtract  %3  %15 3
%21 = OpFOrdLessThanEqual  %20  %22 %2
OpSelectionMerge %23 None
OpBranchConditional %21 %24 %25
%24 = OpLabel
OpReturnValue %4
%25 = OpLabel
OpBranch %23
%23 = OpLabel
%29 = OpLoad  %26  %27
%34 = OpLoad  %31  %32
%39 = OpCompositeExtract  %3  %15 0
%40 = OpCompositeExtract  %3  %15 1
%41 = OpCompositeConstruct  %35  %39 %40
%42 = OpCompositeConstruct  %35  %5 %6
%38 = OpFMul  %35  %41 %42
%44 = OpCompositeExtract  %3  %15 3
%43 = OpFDiv  %3  %4 %44
%37 = OpVectorTimesScalar  %35  %38 %43
%45 = OpCompositeConstruct  %35  %5 %5
%36 = OpFAdd  %35  %37 %45
%46 = OpCompositeExtract  %3  %36 0
%47 = OpCompositeExtract  %3  %36 1
%50 = OpBitcast  %49  %14
%48 = OpConvertUToF  %3  %50
%51 = OpCompositeConstruct  %9  %46 %47 %48
%52 = OpSampledImage  %30  %29 %34
%55 = OpCompositeExtract  %3  %15 2
%57 = OpCompositeExtract  %3  %15 3
%56 = OpFDiv  %3  %4 %57
%54 = OpFMul  %3  %55 %56
%53 = OpImageSampleDrefExplicitLod  %3  %52 %51 %54 Lod %58
OpReturnValue %53
OpFunctionEnd
%64 = OpFunction  %63  None %65
%66 = OpLabel
%59 = OpVariable  %60  Function %8
%61 = OpVariable  %62  Function %12
OpBranch %67
%67 = OpLabel
OpLoopMerge %68 %70 None
OpBranch %69
%69 = OpLabel
%72 = OpLoad  %11  %61
%75 = OpAccessChain  %79  %77 %80
%73 = OpAccessChain  %81  %75 %82
%83 = OpLoad  %11  %73
%84 = OpExtInst  %11  %1 UMin %83 %10
%71 = OpUGreaterThanEqual  %20  %72 %84
OpSelectionMerge %85 None
OpBranchConditional %71 %86 %87
%86 = OpLabel
OpBranch %68
%87 = OpLabel
OpBranch %85
%85 = OpLabel
%89 = OpLoad  %9  %59
%93 = OpLoad  %11  %61
%100 = OpAccessChain  %104  %102 %105
%106 = OpLoad  %11  %61
%98 = OpAccessChain  %107  %100 %106
%96 = OpAccessChain  %108  %98 %109
%110 = OpLoad  %95  %96
%113 = OpLoad  %16  %111
%94 = OpMatrixTimesVector  %16  %110 %113
%92 = OpFunctionCall  %3  %17 %93 %94
%116 = OpLoad  %9  %114
%117 = OpExtInst  %9  %1 Normalize %116
%122 = OpAccessChain  %123  %102 %124
%125 = OpLoad  %11  %61
%121 = OpAccessChain  %126  %122 %125
%120 = OpAccessChain  %127  %121 %128
%119 = OpAccessChain  %129  %120 %130
%131 = OpLoad  %3  %119
%135 = OpAccessChain  %136  %102 %137
%138 = OpLoad  %11  %61
%134 = OpAccessChain  %139  %135 %138
%133 = OpAccessChain  %140  %134 %141
%132 = OpAccessChain  %142  %133 %143
%144 = OpLoad  %3  %132
%148 = OpAccessChain  %149  %102 %150
%151 = OpLoad  %11  %61
%147 = OpAccessChain  %152  %148 %151
%146 = OpAccessChain  %153  %147 %154
%145 = OpAccessChain  %155  %146 %156
%157 = OpLoad  %3  %145
%158 = OpCompositeConstruct  %9  %131 %144 %157
%159 = OpAccessChain  %160  %111 %161
%162 = OpLoad  %3  %159
%163 = OpAccessChain  %164  %111 %165
%166 = OpLoad  %3  %163
%167 = OpAccessChain  %168  %111 %169
%170 = OpLoad  %3  %167
%171 = OpCompositeConstruct  %9  %162 %166 %170
%118 = OpFSub  %9  %158 %171
%172 = OpExtInst  %9  %1 Normalize %118
%173 = OpDot  %3  %117 %172
%174 = OpExtInst  %3  %1 FMax %2 %173
%91 = OpFMul  %3  %92 %174
%178 = OpAccessChain  %179  %102 %180
%181 = OpLoad  %11  %61
%177 = OpAccessChain  %182  %178 %181
%176 = OpAccessChain  %183  %177 %184
%175 = OpAccessChain  %185  %176 %186
%187 = OpLoad  %3  %175
%191 = OpAccessChain  %192  %102 %193
%194 = OpLoad  %11  %61
%190 = OpAccessChain  %195  %191 %194
%189 = OpAccessChain  %196  %190 %197
%188 = OpAccessChain  %198  %189 %199
%200 = OpLoad  %3  %188
%204 = OpAccessChain  %205  %102 %206
%207 = OpLoad  %11  %61
%203 = OpAccessChain  %208  %204 %207
%202 = OpAccessChain  %209  %203 %210
%201 = OpAccessChain  %211  %202 %212
%213 = OpLoad  %3  %201
%214 = OpCompositeConstruct  %9  %187 %200 %213
%90 = OpVectorTimesScalar  %9  %214 %91
%88 = OpFAdd  %9  %89 %90
OpStore %59 %88
OpBranch %70
%70 = OpLabel
%216 = OpLoad  %11  %61
%215 = OpIAdd  %11  %216 %13
OpStore %61 %215
OpBranch %67
%68 = OpLabel
%219 = OpLoad  %9  %59
%220 = OpCompositeConstruct  %16  %219 %4
OpStore %217 %220
OpReturn
OpFunctionEnd
